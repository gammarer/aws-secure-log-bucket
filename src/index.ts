import { SecureBucket, SecureBucketEncryption } from '@yicr/aws-secure-bucket';
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface SecureLogBucketProps {
  readonly bucketName?: string;
  readonly changeClassTransition?: StorageClassTransitionProperty;
}

export interface StorageClassTransitionProperty {
  readonly infrequentAccessDays: number;
  readonly intelligentTieringDays: number;
  readonly glacierDays: number;
  readonly deepArchiveDays: number;
}

export class SecureLogBucket extends SecureBucket {

  constructor(scope: Construct, id: string, props?: SecureLogBucketProps) {
    super(scope, id, {
      bucketName: props?.bucketName,
      encryption: SecureBucketEncryption.KMS_MANAGED,
    });

    // ðŸ‘‡Add Lifecycle rule.
    this.addLifecycleRule({
      id: 'ArchiveStepLifeCycle',
      enabled: true,
      transitions: [
        {
          storageClass: s3.StorageClass.INFREQUENT_ACCESS,
          transitionAfter: cdk.Duration.days(props?.changeClassTransition?.infrequentAccessDays ?? 60),
        },
        {
          storageClass: s3.StorageClass.INTELLIGENT_TIERING,
          transitionAfter: cdk.Duration.days(props?.changeClassTransition?.intelligentTieringDays ?? 120),
        },
        {
          storageClass: s3.StorageClass.GLACIER,
          transitionAfter: cdk.Duration.days(props?.changeClassTransition?.glacierDays ?? 180),
        },
        {
          storageClass: s3.StorageClass.DEEP_ARCHIVE,
          transitionAfter: cdk.Duration.days(props?.changeClassTransition?.deepArchiveDays ?? 360),
        },
      ],
    });
  }
}