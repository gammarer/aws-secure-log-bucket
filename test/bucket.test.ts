import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { SecureLogBucket } from '../src';

describe('SecureLogBucket Testing', () => {

  describe('Normal SecureLogBucket Testing', () => {
    const stack = new Stack(new App(), 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const bucket = new SecureLogBucket(stack, 'SecureLogBucket', {
      bucketName: 'example-log-bucket',
    });

    it('Is Bucket', () => {
      expect(bucket).toBeInstanceOf(s3.Bucket);
    });

    const template = Template.fromStack(stack);

    it('Should have encryption', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        BucketEncryption: Match.objectEquals({
          ServerSideEncryptionConfiguration: [
            {
              ServerSideEncryptionByDefault: {
                SSEAlgorithm: 'aws:kms',
              },
            },
          ],
        }),
      });
    });

    it('Should versioning enabled', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        VersioningConfiguration: Match.objectEquals({
          Status: 'Enabled',
        }),
      });
    });

    it('Should exist lifecycle', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        LifecycleConfiguration: {
          Rules: Match.arrayEquals([
            Match.objectEquals({
              Id: 'ArchiveStepLifeCycle',
              Status: 'Enabled',
              Transitions: Match.arrayEquals([
                Match.objectEquals({
                  StorageClass: 'STANDARD_IA',
                  TransitionInDays: 60,
                }),
                Match.objectEquals({
                  StorageClass: 'INTELLIGENT_TIERING',
                  TransitionInDays: 120,
                }),
                Match.objectEquals({
                  StorageClass: 'GLACIER',
                  TransitionInDays: 180,
                }),
                Match.objectEquals({
                  StorageClass: 'DEEP_ARCHIVE',
                  TransitionInDays: 360,
                }),
              ]),
            }),
          ]),
        },
      });
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('normal-secure-log-bucket');
    });
  });

  describe('Full Props SecureLogBucket Testing', () => {
    const stack = new Stack(new App(), 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const bucket = new SecureLogBucket(stack, 'SecureLogBucket', {
      bucketName: 'example-log-bucket',
      changeClassTransition: {
        infrequentAccessDays: 20,
        intelligentTieringDays: 40,
        glacierDays: 60,
        deepArchiveDays: 80,
      },
    });

    it('Is Bucket', () => {
      expect(bucket).toBeInstanceOf(s3.Bucket);
    });

    const template = Template.fromStack(stack);

    it('Should match lifecycle', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        LifecycleConfiguration: {
          Rules: Match.arrayEquals([
            Match.objectEquals({
              Id: 'ArchiveStepLifeCycle',
              Status: 'Enabled',
              Transitions: Match.arrayEquals([
                Match.objectEquals({
                  StorageClass: 'STANDARD_IA',
                  TransitionInDays: 20,
                }),
                Match.objectEquals({
                  StorageClass: 'INTELLIGENT_TIERING',
                  TransitionInDays: 40,
                }),
                Match.objectEquals({
                  StorageClass: 'GLACIER',
                  TransitionInDays: 60,
                }),
                Match.objectEquals({
                  StorageClass: 'DEEP_ARCHIVE',
                  TransitionInDays: 80,
                }),
              ]),
            }),
          ]),
        },
      });
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('full-props-secure-log-bucket');
    });

  });

  describe('Empty Props SecureLogBucket Testing', () => {
    const stack = new Stack(new App(), 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const bucket = new SecureLogBucket(stack, 'SecureLogBucket');

    it('Is Bucket', () => {
      expect(bucket).toBeInstanceOf(s3.Bucket);
    });

    const template = Template.fromStack(stack);

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('empty-props-secure-log-bucket');
    });

  });

});
