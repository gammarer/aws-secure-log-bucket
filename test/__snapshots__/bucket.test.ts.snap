// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureLogBucket Testing Empty Props SecureLogBucket Testing Should match snapshot: empty-props-secure-log-bucket 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SecureLogBucket2F892D14": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Id": "archive-step-lifecycle-rule",
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 60,
                },
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 120,
                },
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 180,
                },
                Object {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 360,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureLogBucketPolicyD3831FE7": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "SecureLogBucket2F892D14",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SecureLogBucket2F892D14",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SecureLogBucket2F892D14",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SecureLogBucket Testing Full Props SecureLogBucket Testing Should match snapshot: full-props-secure-log-bucket 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SecureLogBucket2F892D14": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "example-log-bucket",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Id": "archive-step-lifecycle-rule",
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 20,
                },
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 40,
                },
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 60,
                },
                Object {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 80,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureLogBucketPolicyD3831FE7": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "SecureLogBucket2F892D14",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SecureLogBucket2F892D14",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SecureLogBucket2F892D14",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SecureLogBucket Testing Normal SecureLogBucket Testing Should match snapshot: normal-secure-log-bucket 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SecureLogBucket2F892D14": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "example-log-bucket",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Id": "archive-step-lifecycle-rule",
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 60,
                },
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 120,
                },
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 180,
                },
                Object {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 360,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureLogBucketPolicyD3831FE7": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "SecureLogBucket2F892D14",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SecureLogBucket2F892D14",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SecureLogBucket2F892D14",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
